ldes-server-infra-mongo/mongo-retention-repository/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/retention/MemberPropertiesRepositoryImpl.java

```diff
@@ -101,4 +112,85 @@ public void removeMemberPropertiesOfCollection(String collectionName) {
 	public void deleteById(String id) {
 		memberPropertiesEntityRepository.deleteById(id);
 	}
+
+	@Override
+	public Stream<MemberProperties> findExpiredMemberProperties(ViewName viewName,
+																TimeBasedRetentionPolicy policy) {
+		return memberPropertiesEntityRepository
+				.findMemberPropertiesEntitiesByCollectionNameAndViewsContainingAndTimestampBefore(
+						viewName.getCollectionName(),
+						viewName.asString(),
+						LocalDateTime.now().minus(policy.duration())
+				)
+				.map(memberPropertiesEntityMapper::toMemberProperties);
+	}
+
+	@Override
+	public Stream<MemberProperties> findExpiredMemberProperties(ViewName viewName,
+																VersionBasedRetentionPolicy policy) {
+		int versionsToKeep = policy.numberOfMembersToKeep();
+		String collectionName = viewName.getCollectionName();
+		String viewNameAsString = viewName.asString();
+
+		final SortOperation sort = sortTimestampDesc();
+		final MatchOperation match =
+				match(Criteria.where(COLLECTION).is(collectionName).and(VIEWS).in(viewNameAsString));
+		final GroupOperation group = groupOnVersionOf();
+		final ProjectionOperation project = projectVersionsToKeep(versionsToKeep);
+		final UnwindOperation unwind = unwindDocuments();
+		final ReplaceRootOperation replaceRoot = getReplaceRootDocuments();
+
+		final Aggregation aggregation = createAggregation(
+				sort, match, group, project, unwind, replaceRoot
+		);
+
+		return mongoTemplate.aggregateStream(aggregation, MemberPropertiesEntity.NAME, MemberProperties.class);
+	}
+
+	@Override
+	public Stream<MemberProperties> findExpiredMemberProperties(ViewName viewName,
+																TimeAndVersionBasedRetentionPolicy policy) {
+		final MatchOperation match =
+				match(Criteria
+						.where(COLLECTION).is(viewName.getCollectionName())
+						.and(VIEWS).in(viewName.asString())
+						.and(TIMESTAMP).lt(LocalDateTime.now().minus(policy.duration())));
```

Comment:
Does this operation differ so much from the hibernate approach?

Reply:
The 'match' part is exactly the same as the hibernate apporach but this is only a small part of the aggregation pipeline below.. This is the difference between timeandversionbased vs versionbased